kotlinx-coroutines-core/common/src/flow/internal/FlowScope.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.flow.internal
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.internal.*
+import kotlinx.coroutines.intrinsics.*
+import kotlin.coroutines.*
+import kotlin.coroutines.intrinsics.*
+
+/**
+ * Creates a flow that also provides a [CoroutineScope] for each collector.
+ *
+ * Provided [CoroutineScope] is similar to [coroutineScope] with the only exception that it *ties* lifecycle of children
+ * and itself regarding the cancellation, thus being cancelled when one of the children becomes cancelled.
+ * For example:
+ * ```
+ * scopedFlow {
+ *     launch {
+ *         throw CancellationException()
+ *     }
+ * } // <- CE will be rethrown here
+ * ```
+ *
+ * Basically, it is shorthand for:
+ * ```
+ * flow {
+ *     coroutineScope {
+ *         ...
+ *     }
+ * }
+ * ```
+ * with additional constraint on cancellation.
+ * To cancel child without cancelling itself, `cancel(ChildCancelledException())` should be used.
+ */
+internal fun <R> scopedFlow(@BuilderInference block: suspend FlowScope<Unit, R>.() -> Unit): Flow<R> =
```

Comment:
More correctly it should be called `unsafeScopedFlow`. I would also consider an alternative signature of this function passing collector as value parameter, so that there is no need to go through delegated call on each emit:

```
fun <T> unsafeScopedFlow(@BuilderInference block: suspend CoroutineScope.(FlowCollector<T>) -> Unit): Flow<T> = 
    unsafeFlow { 
        val collector = this
        flowScope { block(collector) }
    }
```

Reply:
I'd prefer to avoid "unsafe" in operators implementation